fuzz: replace every fuzzer-controlled while loop with a macro

apoelstra opened the pull request:
Limits the number of iterations to 1000 rather than letting the fuzzer do millions or billions of iterations on a single core.

apoelstra commented:
There are several loops in the fuzz test where the number of iterations is directly controlled by the fuzzer, which oftentimes will go for millions of iterations, leading to certain fuzz tests taking forever on a single core. This PR limits these fuzzer-controlled loops to 1000 iterations, an arbitrary high number.

IMHO letting the fuzzer control the number of iterations directly like this, with no restrictions, is a bad idea because the fuzzer will naturally try to hit super high numbers.

I'm looking for a concept ACK (should we limit these) and an approach ACK (is doing this with a macro a reasonable way to go?). Actual ACKs/merges would also be nice :).

apoelstra commented:
On an EPYC 7502 with 256Gb of RAM:

before, -j1
```
real    67m58.883s
user    152m7.845s
sys     4m36.080s
```

after, -j1
```
real    35m57.697s
user    103m10.803s
sys     3m52.910s
```

before, -j32
```
real    60m6.374s
user    153m29.631s
sys     5m14.991s
```

after, -j32
```
real    28m54.078s
user    102m52.337s
sys     4m16.605s
```

MarcoFalke commented:
Concept ACK. @DrahtBot please list the conflicts for the alternative solution that has been proposed.

MarcoFalke commented:
Ok, that didn't work. Basically, it would be nice to be able to still overwrite the limit per fuzz-case:

* #22454: 300
* #22453: 3000
* #22444: 3000
* ...

apoelstra commented:
@MarcoFalke I added a #define variable and capped the `tx_pool` test to 300 iterations to demonstrate how to use it.

MarcoFalke commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -142,7 +144,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
     };
 
-    while (fuzzed_data_provider.ConsumeBool()) {
+    FUZZER_CONTROLLED_LOOP(fuzzed_data_provider) {
```
it would be nice to be able to pass the limit to the loop (in the same line). Otherwise it will get messy if there are nested loops and one forgets to adjust the limit or the loops are moved without moving the limit, which is in a different line.

apoelstra replied:
Sure -- replaced the extra #define variable with a new macro argument

apoelstra replied:
Maybe we should just drop the "default 1000" one and always explicitly set the maximum number of iterations?

MarcoFalke replied:
Yeah, I wouldn't mind that. An additional `, 1000` doesn't seem too verbose.

MarcoFalke commented on src/test/fuzz/addrman.cpp:
```diff
@@ -44,7 +44,7 @@ FUZZ_TARGET_INIT(addrman, initialize_addrman)
         }
     }
-    while (fuzzed_data_provider.ConsumeBool()) {
+    FUZZER_CONTROLLED_LOOP(fuzzed_data_provider) {
```
```suggestion
    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool) {
```

Since some loops (prefer to?) use `fuzzed_data_provider.remaining_bytes()`, maybe make this configurable as well? Feel free to ignore the naming suggestion.

MarcoFalke replied:
@apoelstra Are you still working on this? If not, I am happy to pick up

apoelstra replied:
Hi, sorry, I can pick this up now.

practicalswift commented:
Concept ACK

Nice idea!

apoelstra commented:
Redid the PR. Now it takes the macro from #22649 and just applies it to every single remaining fuzzer loop.

practicalswift commented:
Thanks for fixing this!

Is the mixing of `LIMITED_WHILE (` and `LIMITED_WHILE(` intentional? (Personally I prefer the latter.)

ACK modulo spacing consistency :)

laanwj commented:
> There are several loops in the fuzz test where the number of iterations is directly controlled by the fuzzer, which oftentimes will go for millions of iterations, leading to certain fuzz tests taking forever on a single core. 

Concept and code review ACK be655e8b8111f3a7cd7909203c2bac3d16751406

I would slightly prefer this "arbitrary high number" to be a constant as it's repeated everywhere.

MarcoFalke commented:
> I would slightly prefer this "arbitrary high number" to be a constant as it's repeated everywhere.

I'd say the number is dependent on the test context. The current ones in master are 30, 300, 3000, and 30000.

I think each number needs to be picked carefully to be in the right 10^x.

> Is the mixing of LIMITED_WHILE ( and LIMITED_WHILE( intentional? (Personally I prefer the latter.)

clang-format(-diff) should take care of those in any case.

apoelstra commented:
Oops, I will fix the spacing inconsistency. It probably happened when I switched from manual edits to using sed.

@MarcoFalke I think we could probably start with an "anbitrary high number", say 30000 (or 10000), and then when when tests reveal that certain tests are taking too long we can reduce them in individual cases. The current situation effectively has a limit of infinity and things are mostly OK, so I don' t think this'll be a big deal.

Once we're at 5-digit numbers I don't think there are any situations where we'd want to *increase* them.

MarcoFalke commented:
Sorry, will review soon

apoelstra committed:
fuzz: replace every fuzzer-controlled loop with a LIMITED_WHILE loop

Blindly chose a cap of 10000 iterations for every loop, except for
the two in script_ops.cpp and scriptnum_ops.cpp which appeared to
(sometimes) be deserializing individual bytes; capped those to one
million to ensure that sometimes we try working with massive scripts.

There was also one fuzzer-controlled loop in timedata.cpp which was
already capped, so I left that alone.

git grep 'while (fuzz' should now run clean except for timedata.cpp

MarcoFalke commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -199,7 +199,7 @@ FUZZ_TARGET_INIT(versionbits, initialize)
 
     // mine prior periods
-    while (fuzzed_data_provider.remaining_bytes() > 0) {
+    LIMITED_WHILE(fuzzed_data_provider.remaining_bytes() > 0, 10000) {
```
Why is this needed? There is already a `break` below, no?

apoelstra replied:
Yep, you're right, I misread the loop's exit condition.

MarcoFalke commented on src/test/fuzz/rpc.cpp:
```diff
@@ -348,7 +348,7 @@ FUZZ_TARGET_INIT(rpc, initialize_rpc)
     }
     std::vector<std::string> arguments;
-    while (fuzzed_data_provider.ConsumeBool()) {
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
```
This seems a bit high? `ConsumeRPCArgument` might call `ConsumeArrayRPCArgument`, so the limit is 100'000'000. A user shouldn't be passing in more than 100 items. So any value between 100-1000 seems more appropriate.

See also https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40648

Which is running base58 encoding (expensive) of one rpc arg 6000 times, thus times out. Your limit of 10000 wouldn't affect that, I think.

apoelstra replied:
ACK. I reduced it to 100.

I worry that reducing this will meaningfully affect test coverage since we presumably want to test what happens when you give gazillions of RPC arguments ... but probably a general-purpose fuzz test is the wrong place to try to hit that.

MarcoFalke replied:
Sorry, I also meant in the line above in this file. 100 * 100 is still 10k.

maflcko reviewed (approved):
ACK, beside comments

MarcoFalke commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -199,7 +199,7 @@ FUZZ_TARGET_INIT(versionbits, initialize)
 
     // mine prior periods
-    while (fuzzed_data_provider.remaining_bytes() > 0) {
+    while(fuzzed_data_provider.remaining_bytes() > 0) { // early exit; no need for LIMITED_WHILE
```
```suggestion
    while (fuzzed_data_provider.remaining_bytes() > 0) { // early exit; no need for LIMITED_WHILE
```

nit

apoelstra replied:
fixed

MarcoFalke commented on src/test/fuzz/rpc.cpp:
```diff
@@ -294,7 +294,7 @@ std::string ConsumeScalarRPCArgument(FuzzedDataProvider& fuzzed_data_provider)
 {
     std::vector<std::string> scalar_arguments;
-    while (fuzzed_data_provider.ConsumeBool()) {
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
```
This one I meant, too ;)

MarcoFalke replied:
https://github.com/bitcoin/bitcoin/pull/22508#discussion_r748396777

MarcoFalke commented:
Some times:

master | this | relative | name
```
	206.2	76.67	0.3718234724	rpc			
	149.06	157.45	1.056286059	process_messages																				
	27.21	27.19	0.9992649761	addrman																				
	24.19	24.89	1.028937578	script_sign																				
	23.93	26.08	1.089845382	process_message																				
	23.32	20.44	0.8765008576	coins_view																				
	15.22	15.75	1.034822602	policy_estimator																				
	14.46	15.32	1.059474412	script_ops																				
	13.06	12.85	0.9839203675	bloom_filter																				
	12.35	12.41	1.0048583	pow																				
	11.93	8.79	0.7367979883	connman																				
	10.71	9.38	0.8758169935	net																				
	9.22	7.21	0.7819956616	node_eviction																				
	7.7	8.23	1.068831169	rbf

MarcoFalke commented:
So yeah, this should fix https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=40648

MarcoFalke commented:
cr ACK 214d9055acdd72189a2f415477ce472ca8db4191
